File: trees/level_order_traversal.py
from collections import deque


class TreeNode:
    def __init__(self, value):
        self.val = value
        self.left = None
        self.right = None


def level_order_traversal(node):
    if node is None:
        return

    queue = deque()
    queue.append(node)

    while len(queue) > 0:
        level_size = len(queue)

        for _ in range(level_size):
            curr = queue.popleft()
            print(curr.val, end=" ")

            if curr.left:
                queue.append(curr.left)
            if curr.right:
                queue.append(curr.right)

        print("")


# Example usage:
# Constructing a binary tree
root = TreeNode(1)
root.left = TreeNode(2)
root.right = TreeNode(3)
root.left.left = TreeNode(4)
root.left.right = TreeNode(5)
root.right.left = TreeNode(6)
root.right.right = TreeNode(7)

level_order_traversal(root)

File: trees/levelOrderTraversal.js
/**
 * Performs level order traversal of a binary tree.
 * @param {TreeNode} root - The root node of the binary tree.
 * @return {number[][]} - Array of levels with node values.
 */
function levelOrder(root) {
    if (!root) return []; // If tree is empty, return empty array
    const result = []; // Array to store the level order traversal
    const queue = [root]; // Initialize queue with root node

    while (queue.length) {
        const level = []; // Array to store current level's node values
        const size = queue.length; // Number of nodes at current level

        for (let i = 0; i < size; i++) {
            const node = queue.shift(); // Dequeue the front node
            level.push(node.val); // Add node's value to current level

            // Enqueue left child if exists
            if (node.left) queue.push(node.left);
            // Enqueue right child if exists
            if (node.right) queue.push(node.right);
        }

        result.push(level); // Add current level to result
    }

    return result; // Return the level order traversal
}

// Example Usage:
// Constructing the binary tree:
//     3
//    / \
//   9  20
//      / \
//     15  7
const rootLevel = new TreeNode(3);
rootLevel.left = new TreeNode(9);
rootLevel.right = new TreeNode(20, new TreeNode(15), new TreeNode(7));

console.log(levelOrder(rootLevel)); 
// Output: [[3], [9, 20], [15, 7]]